src/Castle.Core/DynamicProxy/Contributors/InterfaceProxyWithoutTargetContributor.cs

```diff
@@ -62,6 +63,15 @@ protected override IEnumerable<MembersCollector> CollectElementsToProxyInternal(
 
 		private Type GetInvocationType(MetaMethod method, ClassEmitter emitter)
 		{
+			var methodInfo = method.Method;
+
+			if (canChangeTarget == false && methodInfo.IsAbstract && methodInfo.IsGenericMethod == false && methodInfo.IsGenericMethodDefinition == false)
+			{
+				// We do not need to generate a custom invocation type because no custom implementation
+				// for `InvokeMethodOnTarget` will be needed (proceeding to target isn't possible here):
+				return typeof(InterfaceMethodWithoutTargetInvocation);
```

Comment:
I think it would be good to have at least one unit test that verifies this, since this return type is now part of the public API and to verify the performance improvement is working into the future.

I can also see more performance optimisations on the horizon doing better invocation type reuse, so verifying that will also be important.

Reply:
I'll see what I can come up with.

(Speaking briefly about future optimizations: I believe it would be entirely feasible to completely stop generating invocation types, in favor of generating a sibling `InvokeOnTarget` method for each proxied method... directly in the proxy type containing the proxied method. We could then have just one single sealed `Invocation` type that receives a delegate to the appropriate `InvokeOnTarget` method. While that would imply a need to allocate delegates, I suspect this would still be a net win&ndash;in the time that it takes to emit a dynamic type, you can probably do a thousand delegate calls.)

Reply:
I have added a single unit test (for now...more to come if we introduce more invocation type reusing). I could have added more tests, but thought it would make most sense to only write tests for cases where specific invocation types get chosen (there's currently only one such case), rather than also testing when they do _not_ get chosen. The latter could easily be misunderstood to be a specification of desired behavior, when in actual fact it would only document the status quo.

Reply:
> thought it would make most sense to only write tests for cases where specific invocation types get chosen, rather than also testing when they do _not_ get chosen

Completely agree.

> I also refrained from writing any timing-related tests as those can easily become fragile.

Also agree. If we want to do timing tests we should have a separate benchmarks project and use BenchmarkDotNet or something, they can't be run on cloud hosted build servers anyway.